Set up SwiftLint as a standalone build plugin

mokagio opened the pull request:
<!--
- Cannot us the established 0.54.0 version because the plugin starts from 0.55.1
- Using the latest version, 0.58.2, results in SwiftLint taking ~30-40s to run, whether from terminal or within Xcode. Seems similar to what described in https://github.com/realm/SwiftLint/issues/5207 —
- From my quick experiments, earlier version don't support (0.55.x) or have issues with (0.56.x and 0.57.x) the `--working-directory` option which we depend upon

The 0.58.2 performance issue is present in the Linter agent, too. See [this build](https://buildkite.com/automattic/wordpress-ios/builds/25470#0194881f-0a78-4905-beca-cefed3e82132) where it run in 29s:

<img width="688" alt="image" src="https://github.com/user-attachments/assets/6966cd90-de8b-48b8-8785-2042279d487c" />

[Example](https://buildkite.com/automattic/wordpress-ios/builds/25471#0194882d-3336-4cea-a8d7-ddbb4dfcafd4) of the time it takes in CI with this setup:

![image](https://github.com/user-attachments/assets/89237032-2741-47be-ad7d-aa6e7eaf426d)
-->

#23958 removes CocoaPods from the project setup, but with SwiftLint also goes away.

This PR re-introduces SwiftLint, fetching it using the SwiftPM plugin repo but then calling the binary directly via the existing SwiftLint aggregate target (which calls `rake lint` under the hood.

I considered using the plugin setup, see #23990, but didn't go ahead with it because:

- It requires integrating in _every_ target
- It does not seem to care about our configuration
- It requires extract flags for permissions when running in CI and will prompt the user on first run locally
- The setup does not lend itself to having a single source of truth for the SwiftLint version to use, which is a requirement for our CI linter to run—and I strongly advice keeping the CI linter setup as it is
- In my few minutes of tinkering with it, Xcode crashed at least 4 times (see example in #23990 description)
- This is a bit of an argument from authority, which is why I'm leaving it last and using it more as a reinforcement of my conclusion than an explanation, but [even the original SwiftLint author recommend against using the plugin](https://github.com/realm/SwiftLint/issues/5207#issuecomment-1810115901)

For reference, this setup is inspired [by SwiftFormat](https://github.com/nicklockwood/SwiftFormat#using-swift-package-manager). 

I considered fetching SwiftLint directly instead of the plugin, but doing so would have required compiling SwiftLint from source to run it, which was very time consuming.

Finally, it crossed my mind to keep the old CocoaPods setup for SwiftLint. I started going down that path in #23998 but stopped once I noticed that the performance using the setup in this PR were comparable to those on `trunk`, see #23999 

Given that removing CocoaPods whether to build the app or for tooling, brings us closer to our goal of no longer depending on Ruby for product-development tasks, I'd rather stick with this Swift-only approach.

> [!NOTE]  
> I'm having issues archiving the app via Fastlane, even though ~~vanilla `xcodebuild`~~ and Xcode work fine. I'm investigating...
>
> ```
> error: module 'WordPressSharedObjC' in AST file 
> '/Users/gio/Developer/a8c/wpios/DerivedData/ModuleCache.noindex/EQUUY9BHSJ5N/WordPressSharedObjC-5G93B85NZ09I.pcm'
> (imported by AST file '/Users/gio/Developer/a8c/wpios/DerivedData/WordPress/Build/Intermediates.noindex/ArchiveIntermediates/WordPress Alpha/PrecompiledHeaders/WordPress-Bridging-Header-swift_1L0UBHDEION2G-clang_EQUUY9BHSJ5N.pch')
> is not defined in any loaded module map file;
> maybe you need to load /Users/gio/Developer/a8c/wpios/DerivedData/WordPress/Build/Intermediates.noindex/ArchiveIntermediates/WordPress Alpha/IntermediateBuildFilesPath/GeneratedModuleMaps-iphoneos/WordPressSharedObjC.modulemap'?
> ```
>
> Apart from being unsure about the error itself, I don't understand is how the changes in this PR might be affecting the build that way...
>
> **Updated** I haven't fixed the issue but I have a workaround that I like and that I think we should consider in order to merge this PR and then move forward with https://github.com/wordpress-mobile/WordPress-iOS/pull/23958
>
> The TL;DR is that the error occurs only when building builds with configuration Release or Release-Alpha, so I told the script that runs SwiftLint to only run on Debug builds. We have CI covering us when building a different configuration because it runs the standalone linter. In fact, I also disabled the SwiftLint build phase in CI, to save time because of that.

## Regression Notes

1. Potential unintended areas of impact – N.A.
2. What I did to test those areas of impact (or what existing automated tests I relied on) – N.A.
3. What automated tests I added (or what prevented me from doing so) – N.A.

---

**PR submission checklist:**

- [x] I have completed the Regression Notes.
- [x] I have considered adding unit tests for my changes. **N.A.**
- [x] I have considered adding accessibility improvements for my changes. **N.A.**
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary. **N.A.**

**UI changes testing checklist:** Not a UI PR.

mokagio committed:
Merge remote and local SwiftLint configs–Plugin has trouble with remote

mokagio committed:
Fix `opening_braces` violation using `swiftlint --fix`

mokagio committed:
Set SDK when running SwiftLint to avoid conflicts

dangermattic commented:
<!--
  0 Errors
  0 Warnings
  1 Message: This PR is still a Draft: some...
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">This PR is still a Draft: some checks will be skipped.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

mokagio commented on WordPress/Classes/Extensions/NSAttributedString+Helpers.swift:
```diff
@@ -33,8 +33,7 @@ public extension NSAttributedString {
                 let animatedImage = image as? AnimatedImage,
-                animatedImage.gifData != nil
-            {
+                animatedImage.gifData != nil {
```
Moving up from version 0.54.0 brought up this and other `opening_brace` violations.

Might be related to the rule having been [rewritten in version 0.55.0](https://github.com/realm/SwiftLint/releases/tag/0.55.0) and then [updated in version 0.57.0](https://github.com/realm/SwiftLint/releases/tag/0.57.0).

mokagio commented on .swiftlint.yml:
```diff
@@ -1,10 +1,11 @@
-swiftlint_version: 0.54.0
+swiftlint_version: 0.58.2
```
We cannot use `0.54.0` because it's not available as a built tool plugin, see https://github.com/SimplyDanny/SwiftLintPlugins/releases

Given that, it makes sense to use the latest available version. But even if we didn't want to:

- 0.58.0 and 0.58.1 have a bug where the Docker image reads the wrong version, so we cannot use them
- 0.55.0 does not have the `--working-directory` option which we need

mokagio committed:
Wrap SwiftLint call from aggretate target in script

This way, we'll be able to add additional checks without messing
up the project file.

mokagio committed:
Do not run build-time SwiftLint when building in CI

mokagio committed:
Only run SwiftLint build phase in Debug builds

mokagio commented on BuildTools/Package.swift:
```diff
@@ -0,0 +1,39 @@
+    name: "BuildTools",
+    platforms: [.macOS(.v10_13)],
+    dependencies: [
+        // .package(url: "https://github.com/SimplyDanny/SwiftLintPlugins", from: "0.58.2"),
+        // .package(url: "https://github.com/Realm/SwiftLint", exact: loadSwiftLintVersion()),
```
```suggestion
```

kean commented on BuildTools/.gitignore:
```diff
@@ -0,0 +1 @@
+.build
```
I'd add it to the .gitignore file in the root directory.

kean commented on BuildTools/Package.swift:
```diff
@@ -0,0 +1,37 @@
+    name: "BuildTools",
+    platforms: [.macOS(.v10_13)],
+    dependencies: [
+        .package(url: "https://github.com/SimplyDanny/SwiftLintPlugins", exact: loadSwiftLintVersion()),
```
Does the version numbers have to be in `.swiftlint.yml`? Are there any other roadblocks for simply defining the version here in the package:

```
.package(url: "https://github.com/SimplyDanny/SwiftLintPlugins", exact: "0.58.2"),
```

mokagio replied:
`.swiftlint.yml` needs to specify the SwiftLint version the configuration expects

https://github.com/Automattic/a8c-ci-toolkit-buildkite-plugin/blob/29064ab1e6e49135f9700d86b7b3279081b9ebc9/bin/run_swiftlint#L23-L27

We don't _have to_ read the version from there here, but I find it nice to have a single source of truth.

We could duplicate the value here, but we _might_ run into some annoying issues if the version defined here diverges from the one in the config so that we end up with CI running a different version than local and giving different feedback.

kean reviewed (approved):
> This is a bit of an argument from authority, which is why I'm leaving it last and using it more as a reinforcement of my conclusion than an explanation, but https://github.com/realm/SwiftLint/issues/5207#issuecomment-1810115901

I concur. It should be used only when the code requires a tool to compile – mainly code generators.

mokagio committed:
Set up SwiftLint as a SwiftPM plugin in dedicated BuildTools pkg

mokagio committed:
Use SwiftLint versin 0.58.2

The previous version, 0.54.0, is not available as a plugin. And of all
the version available via the plugin, this is the only one that actually
works.

mokagio commented:
I thought I'd merged this already 😳 

Apologies for letting this sit in the limbo for so long...

AliSoftware commented on Scripts/BuildPhases/SwiftLint.sh:
```diff
@@ -0,0 +1,23 @@
+fi
+
+# Only run on debug builds.
+# For some reason, running when trying to archive, via CLI, results in a compilation failure.
+#
+# fatal error: module 'WordPressSharedObjC' in AST file '/path/to/DerivedData/ModuleCache.noindex/EQUUY9BHSJ5N/WordPressSharedObjC-5G93B85NZ09I.pcm'
+# (imported by AST file '/Users/gio/Developer/a8c/wpios/DerivedData/WordPress/Build/Intermediates.noindex/ArchiveIntermediates/WordPress Alpha/PrecompiledHeaders/WordPress-Bridging-Header-swift_1L0UBHDEION2G-clang_EQUUY9BHSJ5N.pch')
+# is not defined in any loaded module map file;
+# maybe you need to load '/Users/gio/Developer/a8c/wpios/DerivedData/WordPress/Build/Intermediates.noindex/ArchiveIntermediates/WordPress Alpha/IntermediateBuildFilesPath/GeneratedModuleMaps-iphoneos/WordPressSharedObjC.modulemap'?
+if [ "${CONFIGURATION}" != "Debug" ]; then
```
Nit: if the error is really related to archiving vs just building, as opposed to debug vs release, I think you could have used the `${ACTION}` env var to check for `build` vs `archive`? 🤷
